src/test/java/org/humancellatlas/ingest/security/web/AuthenticationControllerTest.java

```diff
@@ -0,0 +1,218 @@
+
+    @Nested
+    @DisplayName("Registration")
+    class Registration {
+
+        private static final String PATH = "/auth/registration";
+
+        @Test
+        void byAuthenticatedGuest() throws Exception {
+            //given:
+            String subjectId = "cf12881b";
+            UserInfo userInfo = new UserInfo(subjectId, "https://oidc.domain.tld/auth");
+            Authentication authentication = new OpenIdAuthentication(null, userInfo);
+
+            //and:
+            String accountId = "b4912b3";
+            Account persistentAccount = new Account(accountId, subjectId);
+            doReturn(persistentAccount)
+                    .when(accountService)
+                    .register(any(Account.class));
+
+            //when:
+            MvcResult result = webApp
+                    .perform(post(PATH)
+                            .with(authentication(authentication))
+                            .with(csrf()))
+                    .andReturn();
+
+            //then:
+            MockHttpServletResponse response = result.getResponse();
+            assertThat(response.getStatus()).isEqualTo(HttpStatus.OK.value());
+
+            //and:
+            ObjectMapper objectMapper = new ObjectMapper();
+            var resultingAccount = objectMapper.readValue(response.getContentAsString(), Account.class);
+            assertThat(resultingAccount.getId()).isEqualTo(accountId);
+            assertCorrectRegisteredAccount(userInfo);
+
+        }
+
+        private void assertCorrectRegisteredAccount(UserInfo userInfo) {
+            var accountCaptor = ArgumentCaptor.forClass(Account.class);
+            verify(accountService).register(accountCaptor.capture());
+
+            var registeredAccount = accountCaptor.getValue();
+            assertThat(registeredAccount)
+                    .extracting("providerReference")
+                    .containsExactly(userInfo.getSubjectId());
+        }
+
+        @Test
+        @WithMockUser(roles = {"CONTRIBUTOR"})
+        void byRegisteredUser() throws Exception {
+            // expect:
+            webApp
+                    .perform(post(PATH))
+                    .andExpect(status().isForbidden());
+        }
+
+        /*
+        Similar scenario to byRegisteredUser but somehow the Account was either,
+        1) unrecognised and so was treated as an authenticated Guest, or
+        2) Account was erroneously assigned the Guest role.
+        Essentially, we want to handle duplicated subject id in our system.
+         */
+        @Test
+        void byUnrecognisedRegisteredUser() throws Exception {
+            //given:
+            UserInfo userInfo = new UserInfo("cc9a9a1", "https://secure.tld/auth");
+            Authentication authentication = new OpenIdAuthentication(userInfo);
+
+            //and:
+            doThrow(new DuplicateAccount())
+                    .when(accountService)
+                    .register(any(Account.class));
```

Comment:
We should utilise the matcher here to simulate the two scenarios mentioned in the comment above this test rather than for any account, otherwise this doesn't really test anything

Reply:
In this scenario, as far as the controller is concerned, it's only interested in the external behaviour of its collaborator, which is throwing the exception. The check for if the controller passes the correct information to the service is already defined in another test named `byAuthenticatedGuest` in the `assertCorrectRegisteredAccount` method.

Reply:
Ok, I'm a bit confused by the comment. Is an "authenticated Guest" the same as the GuestAccount? If a GuestAccount is supplied to accountService.register(), I don't think a DuplicateAccount() exception would be thrown. Perhaps this test (i.e asserting the error path when attempting to register UnrecognisedRegisteredUsers ) should foremost exist in the AccountServiceTest.

Reply:
The `GuestAccount` is practically an anonymous User that came to us with a valid access token; that token could refer to previously registered User. The setup upon which this can occur is already discussed in the comment in the code.